contrib/win32/win32compat/misc.c

```diff
@@ -551,16 +549,82 @@ w32_chown(const char *pathname, unsigned int owner, unsigned int group)
 	return -1;
 }
 
-static void
+/* Convert a UNIX time into a Windows file time */
+void
 unix_time_to_file_time(ULONG t, LPFILETIME pft)
 {
 	ULONGLONG ull;
-	ull = UInt32x32To64(t, 10000000) + 116444736000000000;
+	ull = UInt32x32To64(t, RATE_DIFF) + EPOCH_DELTA_US;
 
 	pft->dwLowDateTime = (DWORD)ull;
 	pft->dwHighDateTime = (DWORD)(ull >> 32);
 }
 
+/* Convert a Windows file time into a UNIX time_t */
+void
+file_time_to_unix_time(const LPFILETIME pft, time_t * winTime)
+{
+	*winTime = ((long long)pft->dwHighDateTime << 32) + pft->dwLowDateTime;
+	*winTime -= EPOCH_DELTA_US;
+	*winTime /= RATE_DIFF;		 /* Nano to seconds resolution */
+}
+
+static BOOL
+is_root_or_empty(wchar_t * path)
+{
+	wchar_t * path_start;
+	BOOL has_drive_letter_and_colon;
+	int len;
+	if (!path) 
+		return FALSE;
+	len = wcslen(path);
+	if((len > 1) && __ascii_iswalpha(path[0]) && path[1] == L':')
+		path_start = path + 2;
+	else
+		path_start = path;
+	if ((*path_start == L'\0') || ((*path_start == L'\\' || *path_start == L'/' ) && path_start[1] == L'\0'))
+		return TRUE;
+	return FALSE;
+}
+
+static BOOL
+has_executable_extension(wchar_t * path)
+{
+	wchar_t * last_dot;
+	if (!path)
+		return FALSE;
+
+	last_dot = wcsrchr(path, L'.');
+	if (!last_dot)
+		return FALSE;
+	if (_wcsnicmp(last_dot, L".exe", 4) != 0 && _wcsnicmp(last_dot, L".cmd", 4) != 0 &&
+	_wcsnicmp(last_dot, L".bat", 4) != 0 && _wcsnicmp(last_dot, L".com", 4) != 0)
+		return FALSE; 
+	return TRUE;
+}
+
+int
+file_attr_to_st_mode(wchar_t * path, DWORD attributes)
```

Comment:
The change is to replace the usage of stat because stat require some (read) permission. When the user doesn't have the  permission, stat return file_not_found. that is the reason we don't want to use stat.